ekm/local_key_manager.go

```diff
@@ -0,0 +1,282 @@
+				Version: spec.DataVersionElectra,
+				Electra: v,
+			}
+			return km.signer.SignBlindedBeaconBlock(vBlindedBlock, domain, pk)
+		default:
+			return nil, nil, fmt.Errorf("obj type is unknown: %T", obj)
+		}
+
+	case spectypes.DomainVoluntaryExit:
+		data, ok := obj.(*phase0.VoluntaryExit)
+		if !ok {
+			return nil, nil, errors.New("could not cast obj to VoluntaryExit")
+		}
+		return km.signer.SignVoluntaryExit(data, domain, pk)
+	case spectypes.DomainAggregateAndProof:
+		return km.signer.SignAggregateAndProof(obj, domain, pk)
+	case spectypes.DomainSelectionProof:
+		data, ok := obj.(spectypes.SSZUint64)
+		if !ok {
+			return nil, nil, errors.New("could not cast obj to SSZUint64")
+		}
+
+		return km.signer.SignSlot(phase0.Slot(data), domain, pk)
+	case spectypes.DomainRandao:
+		data, ok := obj.(spectypes.SSZUint64)
+		if !ok {
+			return nil, nil, errors.New("could not cast obj to SSZUint64")
+		}
+
+		return km.signer.SignEpoch(phase0.Epoch(data), domain, pk)
+	case spectypes.DomainSyncCommittee:
+		data, ok := obj.(BlockRootWithSlot)
+		if !ok {
+			return nil, nil, errors.New("could not cast obj to BlockRootWithSlot")
+		}
+		return km.signer.SignSyncCommittee(data.BlockRoot[:], domain, pk)
+	case spectypes.DomainSyncCommitteeSelectionProof:
+		data, ok := obj.(*altair.SyncAggregatorSelectionData)
+		if !ok {
+			return nil, nil, errors.New("could not cast obj to SyncAggregatorSelectionData")
+		}
+		return km.signer.SignSyncCommitteeSelectionData(data, domain, pk)
+	case spectypes.DomainContributionAndProof:
+		data, ok := obj.(*altair.ContributionAndProof)
+		if !ok {
+			return nil, nil, errors.New("could not cast obj to ContributionAndProof")
+		}
+		return km.signer.SignSyncCommitteeContributionAndProof(data, domain, pk)
+	case spectypes.DomainApplicationBuilder:
+		var data *api.VersionedValidatorRegistration
+		switch v := obj.(type) {
+		case *eth2apiv1.ValidatorRegistration:
+			data = &api.VersionedValidatorRegistration{
+				Version: spec.BuilderVersionV1,
+				V1:      v,
+			}
+		default:
+			return nil, nil, fmt.Errorf("obj type is unknown: %T", obj)
+		}
+		return km.signer.SignRegistration(data, domain, pk)
+	default:
+		return nil, nil, errors.New("domain unknown")
+	}
+}
+
+func (km *LocalKeyManager) AddShare(encryptedSharePrivKey []byte, sharePubKey phase0.BLSPubKey) error {
+	km.walletLock.Lock()
+	defer km.walletLock.Unlock()
+
+	sharePrivKeyHex, err := km.operatorDecrypter.Decrypt(encryptedSharePrivKey)
+	if err != nil {
+		return ShareDecryptionError(fmt.Errorf("decrypt: %w", err))
+	}
+
+	sharePrivKey := &bls.SecretKey{}
+	if err := sharePrivKey.SetHexString(string(sharePrivKeyHex)); err != nil {
+		return ShareDecryptionError(fmt.Errorf("decode hex: %w", err))
+	}
+
+	if !bytes.Equal(sharePrivKey.GetPublicKey().Serialize(), sharePubKey[:]) {
+		return ShareDecryptionError(errors.New("share private key does not match public key"))
+	}
+
+	acc, err := km.wallet.AccountByPublicKey(string(sharePrivKeyHex))
```

Comment:
Fix incorrect key usage in `AccountByPublicKey`

Instead of passing `string(sharePrivKeyHex)`, which is a private key, use the derived public key:
```go
acc, err := km.wallet.AccountByPublicKey(sharePrivKey.GetPublicKey().SerializeToHexStr())
```

Reply:
great catch, thanks

Reply:
fixed

Reply:
We need to stop using too generic types like `[]byte` to avoid bugs like this

FYI @moshe-blox